Subject: [PATCH] vold: Conditionally remove secdiscard command

 * Some old devices cannot set lockscreen password with 2436e27
 * To remove secdiscard command, set TARGET_REMOVE_SECDISCARD_COMMAND
   flag to true

diff --git a/Android.mk b/Android.mk
index 14bf99a..840d0cf 100644
--- a/Android.mk
+++ b/Android.mk
@@ -121,6 +121,10 @@
   full_src_files += fs/Exfat.cpp
 endif
 
+ifeq ($(TARGET_REMOVE_SECDISCARD_COMMAND),true)
+  vold_cflags += -DREMOVE_SECDISCARD_COMMAND
+endif
+
 include $(CLEAR_VARS)
 
 LOCAL_ADDITIONAL_DEPENDENCIES := $(LOCAL_PATH)/Android.mk
diff --git a/CryptCommandListener.cpp b/CryptCommandListener.cpp
index 5b5a5bb..15869aa 100644
--- a/CryptCommandListener.cpp
+++ b/CryptCommandListener.cpp
@@ -432,10 +432,12 @@
         return sendGenericOkFailOnBool(cli,
                 e4crypt_destroy_user_storage(parseNull(argv[2]), atoi(argv[3]), atoi(argv[4])));
 
+#ifndef REMOVE_SECDISCARD_COMMAND
     } else if (subcommand == "secdiscard") {
         if (!check_argc(cli, subcommand, argc, 3, "<path>")) return 0;
         return sendGenericOkFailOnBool(cli,
                 e4crypt_secdiscard(parseNull(argv[2])));
+#endif
 
     } else {
         dumpArgs(argc, argv, -1);
diff --git a/Ext4Crypt.cpp b/Ext4Crypt.cpp
index dc2e42a..45bf016 100644
--- a/Ext4Crypt.cpp
+++ b/Ext4Crypt.cpp
@@ -698,6 +698,8 @@
     return res;
 }
 
+#ifndef REMOVE_SECDISCARD_COMMAND
 bool e4crypt_secdiscard(const char* path) {
     return android::vold::runSecdiscardSingle(std::string(path));
 }
+#endif
diff --git a/Ext4Crypt.h b/Ext4Crypt.h
index e90167b..14fc6d8 100644
--- a/Ext4Crypt.h
+++ b/Ext4Crypt.h
@@ -38,5 +38,7 @@
 bool e4crypt_prepare_user_storage(const char* volume_uuid, userid_t user_id, int serial, int flags);
 bool e4crypt_destroy_user_storage(const char* volume_uuid, userid_t user_id, int flags);
 
+#ifndef REMOVE_SECDISCARD_COMMAND
 bool e4crypt_secdiscard(const char* path);
+#endif
 __END_DECLS
diff --git a/KeyStorage.cpp b/KeyStorage.cpp
index 20b2391..c8f34d0 100644
--- a/KeyStorage.cpp
+++ b/KeyStorage.cpp
@@ -532,6 +532,7 @@
     return true;
 }
 
+#ifndef REMOVE_SECDISCARD_COMMAND
 bool runSecdiscardSingle(const std::string& file) {
     if (ForkExecvp(
             std::vector<std::string>{kSecdiscardPath, "--",
@@ -541,6 +542,7 @@
     }
     return true;
 }
+#endif
 
 static bool recursiveDeleteKey(const std::string& dir) {
     if (ForkExecvp(std::vector<std::string>{kRmPath, "-rf", dir}) != 0) {
diff --git a/KeyStorage.h b/KeyStorage.h
index 0734f83..649631d 100644
--- a/KeyStorage.h
+++ b/KeyStorage.h
@@ -85,7 +85,9 @@
 // Securely destroy the key stored in the named directory and delete the directory.
 bool destroyKey(const std::string& dir);
 
+#ifndef REMOVE_SECDISCARD_COMMAND
 bool runSecdiscardSingle(const std::string& file);
+#endif
 }  // namespace vold
 }  // namespace android
 
